File: Competitive Programming 3/Section 1.4/Real Life Harder/10415 - Eb Alto Saxophone Player/C++/main.cpp
#include<bits/stdc++.h>

using namespace std;

typedef vector<pair<int,int> > vP;


map<char,vP > ranges={
    make_pair('c',vP{make_pair(2,4),make_pair(7,10)}),
    make_pair('d',vP{make_pair(2,4),make_pair(7,9)}),
    make_pair('e',vP{make_pair(2,4),make_pair(7,8)}),
    make_pair('f',vP{make_pair(2,4),make_pair(7,7)}),
    make_pair('g',vP{make_pair(2,4)}),
    make_pair('a',vP{make_pair(2,3)}),
    make_pair('b',vP{make_pair(2,2)}),
    make_pair('C',vP{make_pair(3,3)}),
    make_pair('D',vP{make_pair(1,4),make_pair(7,9)}),
    make_pair('E',vP{make_pair(1,4),make_pair(7,8)}),
    make_pair('F',vP{make_pair(1,4),make_pair(7,7)}),
    make_pair('G',vP{make_pair(1,4)}),
    make_pair('A',vP{make_pair(1,3)}),
    make_pair('B',vP{make_pair(1,2)})
};
bool notContained(int i, char c){
    if (c=='\0')
        return true;
    for(auto range : ranges[c])
        if (range.first <= i && i <= range.second)
            return false;
    return true;
}

int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    int t;
    cin >> t;
    cin.ignore();
    while(t--){
        char last ='\0';
        vector<int>output(10);
        string line;
        getline(cin,line);
        for (auto c:line){
            for(auto range: ranges[c])
                for(int i = range.first ; i <= range.second ; i ++)
                    if (notContained(i,last))
                        output[i-1]++;
            last = c;
        }
        for(size_t i = 0 ; i < output.size() ; i ++)
            cout << output[i] << (i+1!=output.size()?" ":"");
        cout << "\n";
    }
    return 0;
}

File: Competitive Programming 3/Section 1.4/Real Life Harder/10415 - Eb Alto Saxophone Player/Python/main.py
ranges={
    'c':[(2,4),(7,10)],
    'd':[(2,4),(7,9)],
    'e':[(2,4),(7,8)],
    'f':[(2,4),(7,7)],
    'g':[(2,4)],
    'a':[(2,3)],
    'b':[(2,2)],
    'C':[(3,3)],
    'D':[(1,4),(7,9)],
    'E':[(1,4),(7,8)],
    'F':[(1,4),(7,7)],
    'G':[(1,4)],
    'A':[(1,3)],
    'B':[(1,2)]
}
def notContained(i,c):
    if c == '\0':
        return True
    for Range in ranges[c]:
        if Range[0]<=i and i <= Range[1]:
            return False
    return True

def main():
    for _ in range(int(input())):
        last = '\0'
        output=[0]*10
        for c in input():
            for Range in ranges[c]:
                for i in range(Range[0],Range[1]+1):
                    if notContained(i,last):
                        output[i-1]+=1
            last = c
        print(str(output).replace(',','')[1:-1])
        
main()